File: src/desphere/codecs.py
"""Sample-coding decoders and the capability gate.

``sample_coding`` in a SPHERE header looks like ``pcm``, ``ulaw``, ``alaw``, or
a base coding plus a compression token, e.g. ``pcm,embedded-shorten-v2.00``.

:func:`resolve_codec` is the single place that decides whether we can handle a
file. The design goal (per project intent) is: support the obvious, clearly
documented, lossless path first, and **fail loudly** on everything else so we
never emit a plausible-but-wrong WAV. New variants slot in by registering a
decoder; until then the gate raises a precise error.
"""

from __future__ import annotations

import struct
from typing import Tuple

from . import g711, shorten
from .errors import (
    DesphereError,
    SphereHeaderError,
    UnsupportedCoding,
    UnsupportedFormat,
)
from .sphere import SphereHeader

# Optional Rust accelerator. Only the heavy numeric kernels (shorten decode and
# G.711 expansion) are delegated; the typed validation/error checks stay here, so
# behavior is identical with or without it. PCM byte-reorder stays pure (the
# strided slice is already C-speed). pip install desphere[fast] to enable.
try:
    import desphere_native as _native
except ImportError:  # pragma: no cover - native is optional
    _native = None

# Map NIST sample_byte_format -> endianness of the stored samples.
#   "01" = low byte first  (little-endian)
#   "10" = high byte first  (big-endian)
#   "1"  = single byte      (order irrelevant)
_BYTE_ORDER = {
    "1": "little",
    "01": "little",
    "10": "big",
}


def _to_little_endian(raw: bytes, n_bytes: int, order: str) -> bytes:
    """Return ``raw`` with each ``n_bytes``-wide sample in little-endian order."""
    if n_bytes == 1 or order == "little":
        return raw
    # Reverse byte order within every n_bytes-sized group using strided slice
    # assignment (C-speed, no numpy dependency).
    out = bytearray(len(raw))
    for i in range(n_bytes):
        out[i::n_bytes] = raw[n_bytes - 1 - i::n_bytes]
    return bytes(out)


class PcmCodec:
    """Linear PCM: a lossless byte-order normalization to little-endian WAV.

    Supports 16-bit and 32-bit samples (the overwhelming majority of SPHERE
    corpora). 8-bit and 24-bit are deliberately rejected for now: their sign
    and packing conventions have not been validated against an oracle, and a
    wrong guess would silently corrupt audio.
    """

    name = "pcm"
    supported_n_bytes = (2, 4)

    @classmethod
    def decode(cls, header: SphereHeader, data: bytes) -> Tuple[int, bytes]:
        n_bytes = header.sample_n_bytes
        if n_bytes not in cls.supported_n_bytes:
            raise UnsupportedFormat(
                f"{n_bytes * 8}-bit PCM not supported yet "
                f"(supported: {[n * 8 for n in cls.supported_n_bytes]} bit)"
            )

        order = _BYTE_ORDER.get(header.sample_byte_format)
        if order is None:
            raise UnsupportedFormat(
                f"unrecognized sample_byte_format {header.sample_byte_format!r} "
                "(supported: '1', '01', '10')"
            )

        little = _to_little_endian(data, n_bytes, order)
        return n_bytes * 8, little


class _G711Codec:
    """Base for the two ITU-T G.711 companding laws (8-bit in, 16-bit out)."""

    name = "g711"
    table: list = []

    @classmethod
    def decode(cls, header: SphereHeader, data: bytes) -> Tuple[int, bytes]:
        if header.sample_n_bytes != 1:
            raise UnsupportedFormat(
                f"{cls.name} expects 1-byte samples, got "
                f"sample_n_bytes={header.sample_n_bytes}"
            )
        if _native is not None:
            return 16, bytes(_native.g711_expand(bytes(data), cls.name == "alaw"))
        return 16, g711.expand(data, cls.table)


class UlawCodec(_G711Codec):
    name = "ulaw"
    table = g711.ULAW_TABLE


class AlawCodec(_G711Codec):
    name = "alaw"
    table = g711.ALAW_TABLE


class ShortenCodec:
    """Embedded-shorten (v2) lossless decompression to little-endian PCM.

    Handles 16-bit PCM shorten types, the lossless mu-law mode (type 8, including
    BITSHIFT, expanded to 16-bit PCM via G.711), and QLPC (LPC-predicted) blocks.
    Only unsupported shorten sample types raise a precise error.
    """

    name = "embedded-shorten-v2.00"

    @classmethod
    def decode(cls, header: SphereHeader, data: bytes) -> Tuple[int, bytes]:
        if _native is not None:
            # Native does the heavy decode+expand and returns PCM bytes; we keep
            # the same typed header cross-checks (on the emitted PCM, 2 B/sample).
            try:
                nchan, _is_ulaw, pcm = _native.shorten_to_pcm(bytes(data))
            except ValueError as exc:
                raise DesphereError(str(exc)) from exc
            pcm = bytes(pcm)
            if nchan != header.channel_count:
                raise UnsupportedFormat(
                    f"shorten channel count {nchan} disagrees with SPHERE header "
                    f"channel_count {header.channel_count}"
                )
            expected = header.sample_count * nchan  # in samples
            got = len(pcm) // 2
            if got < expected:
                raise SphereHeaderError(
                    f"shorten stream decoded {got // nchan} samples/channel, "
                    f"but the SPHERE header declares {header.sample_count} "
                    "(stream truncated or QUIT came early)"
                )
            return 16, (pcm[: expected * 2] if got > expected else pcm)

        values, kind, nchan = shorten.decode(data)
        if nchan != header.channel_count:
            raise UnsupportedFormat(
                f"shorten channel count {nchan} disagrees with SPHERE header "
                f"channel_count {header.channel_count}"
            )
        # Cross-check the decoded length against the header, mirroring the PCM
        # path's truncation guard: a stream that QUITs early would otherwise yield
        # a silently-short WAV. Excess (final-block padding) is trimmed.
        expected = header.sample_count * nchan
        if len(values) < expected:
            raise SphereHeaderError(
                f"shorten stream decoded {len(values) // nchan} samples/channel, "
                f"but the SPHERE header declares {header.sample_count} "
                "(stream truncated or QUIT came early)"
            )
        if len(values) > expected:
            values = values[:expected]
        if kind == "ulaw":
            return 16, g711.expand(bytes(values), g711.ULAW_TABLE)
        # kind == "pcm16": signed 16-bit little-endian
        lo, hi = -32768, 32767
        clipped = [lo if v < lo else hi if v > hi else v for v in values]
        return 16, struct.pack("<%dh" % len(clipped), *clipped)


# Registry of base codings we can decode. Compression tokens are gated
# separately so unsupported compressions fail with a clear message.
# Aliases cover the spellings different encoders write into the header.
_BASE_CODECS = {
    "pcm": PcmCodec,
    "ulaw": UlawCodec,
    "mu-law": UlawCodec,
    "mulaw": UlawCodec,
    "alaw": AlawCodec,
    "a-law": AlawCodec,
}
_COMPRESSORS = {
    "embedded-shorten-v2.00": ShortenCodec,
}


def resolve_codec(header: SphereHeader):
    """Return a codec for ``header`` or raise a precise Unsupported* error."""
    tokens = [t.strip().lower() for t in header.sample_coding.split(",")]
    base = tokens[0]
    compression = tokens[1] if len(tokens) > 1 else None

    if compression:
        if compression not in _COMPRESSORS:
            raise UnsupportedCoding(
                f"compressed coding {header.sample_coding!r} not supported yet "
                f"(compression: {compression!r})"
            )
        # Future: return a decoder that decompresses, then applies the base codec.
        return _COMPRESSORS[compression]

    codec = _BASE_CODECS.get(base)
    if codec is None:
        raise UnsupportedCoding(
            f"sample_coding {base!r} not supported yet "
            f"(supported: {sorted(_BASE_CODECS)})"
        )
    return codec

File: rust/src/codecs.rs
//! Sample-coding decoders + capability gate — Rust twin of `src/desphere/codecs.py`.
//!
//! `decode` is the single place that decides whether we can handle a file:
//! support the obvious lossless path, fail loud on anything unvalidated.

use crate::error::DecodeError;
use crate::sphere::SphereHeader;
use crate::{g711, shorten};

/// NIST `sample_byte_format` -> endianness of the stored samples.
fn byte_order(fmt: &str) -> Option<&'static str> {
    match fmt {
        "1" | "01" => Some("little"),
        "10" => Some("big"),
        _ => None,
    }
}

/// Reverse byte order within each `n_bytes`-wide sample for big-endian input.
fn to_little_endian(raw: &[u8], n_bytes: usize, order: &str) -> Vec<u8> {
    if n_bytes == 1 || order == "little" {
        return raw.to_vec();
    }
    let mut out = raw.to_vec();
    for chunk in out.chunks_exact_mut(n_bytes) {
        chunk.reverse();
    }
    out
}

fn pcm_table_to_le(values: &[i16]) -> Vec<u8> {
    let mut out = Vec::with_capacity(values.len() * 2);
    for &v in values {
        out.extend_from_slice(&v.to_le_bytes());
    }
    out
}

/// Decode a payload per its header. Returns `(bits_per_sample, little_endian_pcm)`.
pub fn decode(h: &SphereHeader, data: &[u8]) -> Result<(u16, Vec<u8>), DecodeError> {
    let coding = h.sample_coding();
    let tokens: Vec<String> = coding
        .split(',')
        .map(|t| t.trim().to_ascii_lowercase())
        .collect();
    let base = tokens[0].as_str();
    let compression = if tokens.len() > 1 && !tokens[1].is_empty() {
        Some(tokens[1].as_str())
    } else {
        None
    };

    if let Some(comp) = compression {
        if comp != "embedded-shorten-v2.00" {
            return Err(DecodeError::Unsupported(format!(
                "compressed coding {coding:?} not supported yet (compression: {comp:?})"
            )));
        }
        return decode_shorten(h, data);
    }

    match base {
        "pcm" => decode_pcm(h, data),
        "ulaw" | "mu-law" | "mulaw" => decode_g711(h, data, false),
        "alaw" | "a-law" => decode_g711(h, data, true),
        other => Err(DecodeError::Unsupported(format!(
            "sample_coding {other:?} not supported yet"
        ))),
    }
}

fn decode_pcm(h: &SphereHeader, data: &[u8]) -> Result<(u16, Vec<u8>), DecodeError> {
    let n = h.sample_n_bytes()? as usize;
    if n != 2 && n != 4 {
        return Err(DecodeError::Unsupported(format!(
            "{}-bit PCM not supported yet (supported: 16, 32 bit)",
            n * 8
        )));
    }
    let fmt = h.sample_byte_format();
    let order = byte_order(fmt).ok_or_else(|| {
        DecodeError::Unsupported(format!(
            "unrecognized sample_byte_format {fmt:?} (supported: '1', '01', '10')"
        ))
    })?;
    Ok(((n * 8) as u16, to_little_endian(data, n, order)))
}

/// Expand G.711 companded bytes to little-endian 16-bit PCM. The heavy kernel,
/// shared by the codec and the Python binding (`alaw` selects a-law vs mu-law).
pub fn g711_expand(data: &[u8], alaw: bool) -> Vec<u8> {
    let table = if alaw {
        g711::alaw_table()
    } else {
        g711::ulaw_table()
    };
    let mut out = Vec::with_capacity(data.len() * 2);
    for &b in data {
        out.extend_from_slice(&table[b as usize].to_le_bytes());
    }
    out
}

fn decode_g711(h: &SphereHeader, data: &[u8], alaw: bool) -> Result<(u16, Vec<u8>), DecodeError> {
    if h.sample_n_bytes()? != 1 {
        let name = if alaw { "alaw" } else { "ulaw" };
        return Err(DecodeError::Unsupported(format!(
            "{name} expects 1-byte samples, got sample_n_bytes={}",
            h.sample_n_bytes()?
        )));
    }
    Ok((16, g711_expand(data, alaw)))
}

/// Decode an embedded-shorten stream to little-endian 16-bit PCM, WITHOUT the
/// header cross-checks. Returns `(channel_count, is_ulaw, pcm)`. The heavy
/// kernel, shared by the codec and the Python binding; callers cross-check the
/// length/channels against their header.
pub fn shorten_to_pcm(data: &[u8]) -> Result<(usize, bool, Vec<u8>), DecodeError> {
    let (values, kind, nchan) = shorten::decode(data)?;
    let pcm = match kind {
        shorten::Kind::Ulaw => {
            let table = g711::ulaw_table();
            let expanded: Vec<i16> = values.iter().map(|&v| table[(v as u8) as usize]).collect();
            pcm_table_to_le(&expanded)
        }
        shorten::Kind::Pcm16 => {
            let clipped: Vec<i16> = values
                .iter()
                .map(|&v| v.clamp(-32768, 32767) as i16)
                .collect();
            pcm_table_to_le(&clipped)
        }
    };
    Ok((nchan, kind == shorten::Kind::Ulaw, pcm))
}

fn decode_shorten(h: &SphereHeader, data: &[u8]) -> Result<(u16, Vec<u8>), DecodeError> {
    let (nchan, _is_ulaw, pcm) = shorten_to_pcm(data)?;
    if nchan as i64 != h.channel_count()? {
        return Err(DecodeError::Unsupported(format!(
            "shorten channel count {nchan} disagrees with SPHERE header channel_count {}",
            h.channel_count()?
        )));
    }
    // Cross-check sample count on the emitted PCM (2 bytes/sample).
    let expected = (h.sample_count()? * nchan as i64) as usize * 2;
    if pcm.len() < expected {
        return Err(DecodeError::Corrupt(format!(
            "shorten stream decoded {} samples/channel, but the SPHERE header declares {} (truncated or QUIT came early)",
            pcm.len() / 2 / nchan,
            h.sample_count()?
        )));
    }
    let pcm = if pcm.len() > expected {
        pcm[..expected].to_vec()
    } else {
        pcm
    };
    Ok((16, pcm))
}
